modules/contracts/apps/challenge/src/handlers/execute.rs

```diff
@@ -0,0 +1,568 @@
+    app.admin.assert_admin(deps.as_ref(), &info.sender)?;
+
+    //check that the challenge status is ChallengeStatus::Uninitialized
+    if challenge.status != ChallengeStatus::Uninitialized {
+        return Err(AppError::WrongChallengeStatus {});
+    }
+
+    let mut challenge = challenge.set_end_timestamp(&env)?;
+
+    // Generate the challenge id and update the status
+    let challenge_id = NEXT_ID.update(deps.storage, |id| AppResult::Ok(id + 1))?;
+    challenge.status = ChallengeStatus::Active;
+    CHALLENGE_LIST.save(deps.storage, challenge_id.clone(), &challenge)?;
+
+    // Create the initial check_in entry
+    DAILY_CHECK_INS.save(
+        deps.storage,
+        challenge_id.clone(),
+        &vec![CheckIn::default_from(&env)],
+    )?;
+
+    // Create the initial challenge_votes entry
+    CHALLENGE_VOTES.save(deps.storage, challenge_id.clone(), &Vec::new())?;
+
+    Ok(app.tag_response(
+        Response::new().add_attribute("challenge_id", challenge_id.to_string()),
+        "create_challenge",
+    ))
+}
+
+fn update_challenge(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    app: ChallengeApp,
+    challenge_id: u64,
+    new_challenge: ChallengeEntryUpdate,
+) -> AppResult {
+    app.admin.assert_admin(deps.as_ref(), &info.sender)?;
+
+    // will return an error if the challenge doesn't exist
+    let mut loaded_challenge: ChallengeEntry = CHALLENGE_LIST
+        .may_load(deps.storage, challenge_id.clone())
+        .map_err(|_| AppError::NotFound {})?
+        .ok_or_else(|| {
+            AppError::Std(StdError::generic_err(format!(
+                "Challenge with id {} not found",
+                challenge_id
+            )))
+        })?;
+
+    if loaded_challenge.status != ChallengeStatus::Active {
+        return Err(AppError::WrongChallengeStatus {});
+    }
+
+    if let Some(name) = new_challenge.name {
+        loaded_challenge.name = name;
+    }
+    if let Some(collateral) = new_challenge.collateral {
+        loaded_challenge.collateral = collateral;
+    }
+    if let Some(description) = new_challenge.description {
+        loaded_challenge.description = description;
+    }
+    if let Some(end) = new_challenge.end {
+        loaded_challenge.end = EndType::Duration(DurationChoice::Week);
+    }
+
+    // Save the updated challenge
+    CHALLENGE_LIST.save(deps.storage, challenge_id.clone(), &loaded_challenge)?;
+
+    Ok(app.tag_response(
+        Response::new().add_attribute("challenge_id", challenge_id.to_string()),
+        "update_challenge",
+    ))
+}
+
+fn cancel_challenge(
+    deps: DepsMut,
+    info: MessageInfo,
+    app: &ChallengeApp,
+    challenge_id: u64,
+) -> AppResult {
+    app.admin.assert_admin(deps.as_ref(), &info.sender)?;
+    let mut challenge = CHALLENGE_LIST.load(deps.storage, challenge_id.clone())?;
+    if challenge.status != ChallengeStatus::Active {
+        return Err(AppError::WrongChallengeStatus {});
+    }
```

Comment:
Would we want to be able to cancel a challenge if it is in the over status?

Reply:
I think no, we want the history to persist. Challenge's should only be cancelable if they are active right?